java-checks/src/main/java/org/sonar/java/checks/spring/SpringIncompatibleTransactionalCheck.java

```diff
@@ -0,0 +1,196 @@
+        if (methodsPropagationMap.containsKey(calleeMethodSymbol) && methodInvocationOnThisInstance(methodInvocation)) {
+          String calleePropagation = methodsPropagationMap.get(calleeMethodSymbol);
+          checkIncompatiblePropagation(methodInvocation, callerPropagation, calleeMethodSymbol, calleePropagation);
+        }
+      }
+    });
+  }
+
+  private static boolean methodInvocationOnThisInstance(MethodInvocationTree methodInvocation) {
+    ExpressionTree expression = methodInvocation.methodSelect();
+    if (expression.is(Tree.Kind.MEMBER_SELECT)) {
+      expression = ((MemberSelectExpressionTree) expression).expression();
+      return expression.is(Tree.Kind.IDENTIFIER) && ((IdentifierTree) expression).name().equals("this");
+    }
+    return expression.is(Tree.Kind.IDENTIFIER);
+  }
+
+  private void checkIncompatiblePropagation(MethodInvocationTree methodInvocation, @Nullable String callerPropagation, Symbol calleeMethodSymbol, String calleePropagation) {
+    Set<String> incompatiblePropagation = INCOMPATIBLE_PROPAGATION_MAP.getOrDefault(callerPropagation, Collections.emptySet());
+    if (incompatiblePropagation.contains(calleePropagation)) {
+      String message = "\"" + calleeMethodSymbol.name() + "'s\" @Transactional requirement is incompatible with the one for this method.";
+      List<JavaFileScannerContext.Location> secondaryLocations = Collections.singletonList(
+        new JavaFileScannerContext.Location("", ((MethodTree) calleeMethodSymbol.declaration()).simpleName()));
+      reportIssue(ExpressionUtils.methodName(methodInvocation), message, secondaryLocations, null);
+    }
+  }
+
+  private static Map<Symbol, String> collectMethodsPropagation(ClassTree classTree) {
+    Map<Symbol, String> methodPropagationMap = new HashMap<>();
+    String classPropagation = getPropagation(classTree.symbol(), null);
+    for (Tree member : classTree.members()) {
+      if (member.is(Tree.Kind.METHOD)) {
+        MethodTree method = (MethodTree) member;
+        methodPropagationMap.put(method.symbol(), getPropagation(method.symbol(), classPropagation));
+      }
+    }
+    return methodPropagationMap;
+  }
+
+  private static boolean hasSameValues(Map<Symbol, String> methodsPropagationMap) {
+    if (methodsPropagationMap.size() < 2) {
+      return true;
+    }
+    String firstValue = methodsPropagationMap.values().iterator().next();
+    return methodsPropagationMap.values().stream().allMatch(value -> Objects.equals(value, firstValue));
+  }
+
+  private static boolean hasMethodsAnnotatedWithTransactional(ClassTree classTree) {
+    return classTree.symbol().memberSymbols().stream()
+      .map(Symbol::metadata)
+      .anyMatch(metadata -> metadata.isAnnotatedWith(SPRING_TRANSACTIONAL_ANNOTATION) ||
+        metadata.isAnnotatedWith(JAVAX_TRANSACTIONAL_ANNOTATION));
+  }
+
+  @CheckForNull
+  private static String getPropagation(Symbol symbol, @Nullable String inheritedPropagation) {
```

Comment:
What about inheritance of annotation through class hierarchy ? how does it work in spring and do we support this or not ? please add a testcase to show the support of it or not.

Reply:
I added a test case to show that we don't support class hierarchy